lib/logstash/plugin_mixins/jdbc.rb

```diff
@@ -206,50 +209,58 @@ def prepare_jdbc_connection
   public
   def close_jdbc_connection
     begin
+      # pipeline restarts can also close the jdbc connection, block until the current executing statement is finished to avoid leaking connections
+      # connections in use won't really get closed
+      @connection_lock.lock
       @database.disconnect if @database
     rescue => e
       @logger.warn("Failed to close connection", :exception => e)
+    ensure
+      @connection_lock.unlock
     end
   end
 
   public
   def execute_statement(statement, parameters)
-    success = false
```

Comment:
github diff sucks....here is the real change:
<img width="1769" alt="image" src="https://user-images.githubusercontent.com/976291/33690246-588e6e16-daa8-11e7-93f6-d0e71e565f50.png">